Allow deletion of categories when there are no resources associated

alecslupu opened the pull request:
#### :tophat: What? Why?
This PR adds `dependent: :destroy` to `:categorization` to so that the link between resource and category to be removed. 

#### :pushpin: Related Issues
*Link your PR to an issue*
- Fixes #8389

#### Testing
1. Login as admin 
2. visit  a Participatory Process admin 
3. Add a new category from sidebar 
4. Add a new meeting and assign it to category newly created
5. Visit again the category page and see you cannot delete it 
6. Visit meetings and delete
7. Repeat step 5
8. Apply patch **and restart your server**
9. Visit again the categories and see you can delete your category. 
10. For sanity, repeat 3, 4, 5, 6 then you could jump to step 9 

### :camera: Screenshots
[Screencast from 2023-12-18 18-34-08.webm](https://github.com/decidim/decidim/assets/105683/14560923-3c81-44a1-be99-8e259a407c8a)


:hearts: Thank you!

andreslucena commented:
@alecslupu can you fill the "Testing" section here? I'm trying to follow the instructions that [I've added in the issue](https://github.com/decidim/decidim/issues/8389#issuecomment-966944262) but it doesn't allow me to delete the empty category: 

1. create a process
2. create a category
3. create a meeting assigned to this category
4. delete the meeting
5. try to delete the category (you can't!)

![Screenshot of the Categories section in a process](https://github.com/decidim/decidim/assets/717367/170b1d2c-b919-4b27-a5f8-d1f8714372d5)

alecslupu commented:
> @alecslupu can you fill the "Testing" section here? I'm trying to follow the instructions that [I've added in the issue](https://github.com/decidim/decidim/issues/8389#issuecomment-966944262) but it doesn't allow me to delete the empty category:
> 
>     1. create a process
> 
>     2. create a category
> 
>     3. create a meeting assigned to this category
> 
>     4. delete the meeting
> 
>     5. try to delete the category (you can't!)
> 
> 
> ![Screenshot of the Categories section in a process](https://private-user-images.githubusercontent.com/717367/291322920-170b1d2c-b919-4b27-a5f8-d1f8714372d5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5MTc3MDYsIm5iZiI6MTcwMjkxNzQwNiwicGF0aCI6Ii83MTczNjcvMjkxMzIyOTIwLTE3MGIxZDJjLWI5MTktNGIyNy1hNWY4LWQxZjg3MTQzNzJkNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIxOFQxNjM2NDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MTczNjEyOWI0NzUzOTQwZGFmNDBmMjllM2EyY2NhMmE0MDFmZDU1OTQ4YTFiMjU5OGU5ZGZmMGNhZWMzMjhhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.YTS-1hvmpvE7F4DBmccZ7bODxuZQEfeOk-ZwPHRYX6U)

@andreslucena I have updated the Testing steps and also added a video. Since has_category is a module, maybe you forgot to restart the dev server?

andreslucena commented:
I think its not related to the autoloading, but that it isn't working correctly with editions. For instance, I could remove a Category that was associated to a Meeting:

[screen-recorder-tue-dec-19-2023-13-00-32.webm](https://github.com/decidim/decidim/assets/717367/9daa0ec4-2fed-4418-8eef-a91093b11a95)

alecslupu commented:
Marking as draft for further investigation.

alecslupu commented:
@andreslucena I had a deeper look into this issue, and i found the following: 

- When a resource that has category attached is being destroyed, the categorization was left behind. (ex: When a meeting with associated category is being destroyed, the `Decidim::Categorization` record was left behind. *Fixed* by adding `dependent: :destroy` ) 
- When a subcategoy is being displayed, there was no check to prevent you destroy it if there are associated resources ( fixed by template patch ) 
- I have added also a check in the destroy command, so that we make sure the command is broadcasting invalid if there are any associated resources

andreslucena commented:
@alecslupu let's retake this PR. Can you merge with develop and solve the conflicts?

alecslupu commented:
> @alecslupu let's retake this PR. Can you merge with develop and solve the conflicts?

@andreslucena , the conflicts have been solved.

andreslucena commented on RELEASE_NOTES.md:
```diff
@@ -106,7 +106,17 @@ In case you have modifications in your application's webpack configuration, adap
-### 3.4. [[TITLE OF THE ACTION]]
+### 3.4. Remove orphan categorizations
+
+In order to ensure that the database is consistent, we have removed the orphan categorizations. This is a one time action that needs to be done after the code is updated in the production database.
```
I think that implementers may not know what's an orphan categorization. How about this?

andreslucena commented on RELEASE_NOTES.md:
````diff
@@ -29,7 +29,17 @@ bin/rails db:migrate
+In order to ensure that the database is consistent, we have removed the orphan categorizations. This is a one time action that needs to be done after the code is updated in the production database.
+
+```console
+bin/rails decidim:upgrade:remobe_orphan_categorizations
````
There's a typo here and also I'd prefer to use the `decidim:upgrade:fix_` prefix to be consistent with other tasks:

```console
$ bin/rails -T |grep decidim:upgrade
rails decidim:upgrade                                                                                      # Install migrations from Decidim to the app
rails decidim:upgrade:fix_duplicate_endorsements                                                           # Remove duplicated endorsements
rails decidim:upgrade:fix_nickname_uniqueness                                                              # Modify nicknames with random numbers when exists similar ones case-insensitively
rails decidim:upgrade:fix_short_urls                                                                       # Fix wrongly mapped short links components
rails decidim:upgrade:migrate_wysiwyg_content                                                              # Updates the content entered through the WYSIWYG editors
rails decidim:upgrade:moderation:fix_blocked_user_panel                                                    # Add all blocked users to global moderation panel
rails decidim:upgrade:moderation:remove_from_search                                                        # Exclude all moderated content from search
rails decidim:upgrade:register_wysiwyg_migration                                                           # Registers records for the WYSIWYG content migration
rails decidim:upgrade:remobe_orphan_categorizations                                                        # Remove orphan categorizations
rails decidim:upgrade_app                                                                                  # Applies upgrade modifications to the already installed application
```

andreslucena commented on RELEASE_NOTES.md:
```diff
@@ -106,7 +106,17 @@ In case you have modifications in your application's webpack configuration, adap
 You can read more about this change on PR [\#12238](https://github.com/decidim/decidim/pull/12238).
 
-### 3.4. [[TITLE OF THE ACTION]]
+### 3.4. Remove orphan categorizations
```
```suggestion
### 3.4. Allow removal of orphan categories
```

andreslucena replied:
Without the "Allow" it seems like we're going to delete/destroy categories 😄

andreslucena reviewed (changes requested):
Thanks for all the fixes. I have some suggestions and we're ready to a merge

alecslupu commented on decidim-core/lib/decidim/core/test/shared_examples/has_category.rb:
```diff
@@ -1,11 +1,37 @@
 # frozen_string_literal: true
 
 shared_examples_for "has category" do
+  let(:participatory_space) { subject.participatory_space }
```
I am not 100% what is to be done here?
